staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/client_ca.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dynamiccertificates
+
+import (
+	"reflect"
+)
+
+// CAContentProvider provides ca bundle byte content
+type CAContentProvider interface {
+	// Name is just an identifier
+	Name() string
+	// CurrentCABundleContent provides ca bundle byte content
+	CurrentCABundleContent() []byte
```

Comment:
I suspect we may want to add an error, especially for later cases where this function may involve reading a file. It seems like currently we are using a return of nil to denote a problem reading?

Reply:
> I suspect we may want to add an error, especially for later cases where this function may involve reading a file. It seems like currently we are using a return of nil to denote a problem reading?

I think you have two different controllers.  One indicating failure to read the value and the consumer of this API just gets the latest value.

Reply:
I think the error should be reported at a lower level by the controller doing the read.  By the time we get all the way here, we want the best value possible.